File: spark-join-parity/code/explain_plan.py
from pyspark.sql import SparkSession

# Produces results/physical_plan.txt: prints the exact physical plan
# for a plain PySpark DataFrame join, with the same configs used by
# join_bench.py (forced SortMergeJoin, no AQE). This is the source of
# the "explain(\"formatted\") on both confirms..." claim in the article.
spark = (
    SparkSession.builder
    .master("local[4]")
    .config("spark.driver.host", "127.0.0.1")
    .config("spark.driver.bindAddress", "127.0.0.1")
    .config("spark.sql.shuffle.partitions", "8")
    .config("spark.sql.autoBroadcastJoinThreshold", "-1")
    .config("spark.sql.adaptive.enabled", "false")
    .getOrCreate()
)
spark.sparkContext.setLogLevel("WARN")

events = spark.read.parquet("/tmp/bench/data/events")
accounts = spark.read.parquet("/tmp/bench/data/accounts")

joined = events.join(accounts, on="id", how="inner")
joined.explain("formatted")

spark.stop()

File: spark-join-parity/code/explain_plan.scala
// Scala counterpart to explain_plan.py: prints the physical plan for
// the typed Dataset.joinWith version, same configs as JoinBench.scala.
// Run via spark-shell -i, with the same --conf flags used for the
// timed benchmark (see README.md).
case class Event(id: Long, ts: Long, amount: Double, category: String)
case class Account(id: Long, region: String, tier: Int, signup_epoch: Long)

import spark.implicits._

val events = spark.read.parquet("/tmp/bench/data/events").as[Event]
val accounts = spark.read.parquet("/tmp/bench/data/accounts").as[Account]
val joined = events.joinWith(accounts, events("id") === accounts("id"), "inner")
joined.explain("formatted")

System.exit(0)
